crates/polars-expr/src/expressions/literal.rs

```diff
@@ -101,7 +101,40 @@ impl PhysicalExpr for LiteralExpr {
         state: &ExecutionState,
     ) -> PolarsResult<AggregationContext<'a>> {
         let s = self.evaluate(df, state)?;
-        Ok(AggregationContext::from_literal(s, Cow::Borrowed(groups)))
+
+        if self.0.is_scalar() {
+            Ok(AggregationContext::from_agg_state(
+                AggState::LiteralScalar(s),
+                Cow::Borrowed(groups),
+            ))
+        } else {
+            // A non-scalar literal value expands to those values for every group.
```

Comment:
Non-scalar literals are now lowered to `AggregatedList` with all the values per group. This is a more correct handling of those literals. For example, now:

```python
>>> df = pl.DataFrame(
...     {
...         "g": [0, 1],
...     }
... )
>>> df.group_by("g").agg(foo=pl.lit(pl.Series([42, 2, 3])))
shape: (2, 2)
┌─────┬────────────┐
│ g   ┆ foo        │
│ --- ┆ ---        │
│ i64 ┆ list[i64]  │
╞═════╪════════════╡
│ 0   ┆ [42, 2, 3] │
│ 1   ┆ [42, 2, 3] │
└─────┴────────────┘
```